### DISCUSSION TITLE
[C128] Why are adding Memory Sections & Segment Sections not shown in the %O file?

### DISCUSSION THREAD

**Discussion by RetrofanC** [cc65/cc65#1588] [Q&A] [answered]
Hi,
I am making nice progress with the C128 multi-bank configuration for CC65.
So have been deep into the different CC65 documentation (ar65,ld65,ca65).
Changed the crt0.s, c129.cfg, c128.lib, and using #PRAGMA's and .BANK capabilities, etc.

However, there is one thing I cannot seem to find in the documentation nor the internet forum or old Mailing lists. 

In my new c128.cfg, I have added a COMMON: (free memory area section) in the memory section and a BANKING: segment (in the segment section).  Now, I want this new definitions to be part of the default %O file, but can't make that happen. When I define it as such, I cannot find it anywhere in the a.out, but also no error message is given. When I use `file="BANKING"`, surely it all works, and I can BLOAD it separately, but that is not what I am looking for. Especially as it is the COMMON memory which can be part of the program loaded into the standard cc65 config c128 - RAM 0 memory.

**MEMORY SECTION**
_These work:_
(Meaning the COMMON are not used at the same time, but one to test, the others are # out.)
```ini
MAIN:     file = %O, define = yes, start = $1C0D, size = $A3F0 - __STACKSIZE__, bank = 99;

COMMON:   file = "commonram.prg", start = $0402, size = $03fd, define = yes, bank = 99;
COMMON:   file = "commonram.prg", start = $03E4, size = $001B, define = yes, bank = 99;
COMMON:   file = "commonram.prg", start = $0202, size = $00A2, define = yes, bank = 99;
```
_These individual lines **don't** work;_ 
_(Meaning no error message, but also no code in the %O for those memory locations.)_
_(Meaning they are not used at the same time, but one to test, the others are # out.)_
```ini
COMMON:   file = %O, start = $0402, size = $03fd, define = yes, bank = 99;
COMMON:   file = %O, start = $03E4, size = $001B, define = yes, bank = 99;
COMMON:   file = %O, start = $0202, size = $00A2, define = yes, bank = 99;
```
To make sure it had to do with the specifics of the common RAM in the c128, I also tried to split the existing MAIN:
```ini
MAIN:     file = %O, define = yes, start = $1C0D, size = $83F0 - __STACKSIZE__, bank = 99;
MAIN:     file = %O, define = yes, start = $83F1, size = $A3F3, bank = 98;
```
But, it seems that every 'new' memory section created is not taken into account for creating the %O.

**SEGMENT SECTION**
In the segment section, I am using this definition which works for the `file = "commonram.prg"`, but doesn't do anything for the `file = %O`.
```ini
BANKING:   load = COMMON,   type = ro,  define = yes,  optional = yes;
```
(One other strange thing is that I have to use the argument "optional = yes"; otherwise, I get an error message, and not sure why.)

Any guidance or suggestions are helpful to solve this (or, make clear it can't be done in cc65 toolchain).
I looked also into the "exehdr.s", but that didn't provide any solution direction either unfortunate. 

To conclude, I also looked into the "load = " and "run = " with separate values, but that didn't do the trick either.

Thanks in advance for your answers & insights.

**Comment by mrdudz** [comment]
please give a full example, it should work :)

**Reply by RetrofanC** [reply]
Thanks for the quick reply!

These are the three different versions I have been trying.
I also moved the `COMMON:` definition in the file to be between the `ZP:` and `LOADADDR:` lines.
_(As I thought perhaps it cannot append the file if the memory range isn't sequential.)_

Please let me know if you need more information to determine where I am missing out to make it work.

**c128.cfg - Option `#1`**
_(The first two c128.cfg versions using the 1K common RAM with no change to "crt0.s".)_

```ini
SYMBOLS {
    __LOADADDR__:  type = import;
    __EXEHDR__:    type = import;
    __STACKSIZE__: type = weak, value = $0800; # 2k stack
}
MEMORY {
    ZP:                  file = "", define = yes, start = $0002, size = $001A;
    LOADADDR:   file = %O,               start = $1BFF, size = $0002;
    HEADER:         file = %O,               start = $1C01, size = $000C;
    MAIN:             file = %O, define = yes, start = $1C0D, size = $A3F0 - __STACKSIZE__, bank = 99; #  $A3F3
   COMMON:      file = %O, define = yes, start = $03E4, size = $001B;
}
SEGMENTS {
    ZEROPAGE: load = ZP,       type = zp;
    LOADADDR: load = LOADADDR, type = ro;
    EXEHDR:   load = HEADER,   type = ro;
    STARTUP:  load = MAIN,     type = ro;
    LOWCODE:  load = MAIN,     type = ro,  optional = yes;
    ONCE:     load = MAIN,     type = ro,  optional = yes;
    CODE:     load = MAIN,     type = ro;
    RODATA:   load = MAIN,     type = ro;
    DATA:     load = MAIN,     type = rw;
    INIT:     load = MAIN,     type = bss;
    BSS:      load = MAIN,     type = bss, define   = yes;

#The following segment common memory is added for bank switching code
     BANKING:   load = COMMON,   type = ro,  define = yes,  optional = yes;
}
FEATURES {
    CONDES: type    = constructor,
            label   = __CONSTRUCTOR_TABLE__,
            count   = __CONSTRUCTOR_COUNT__,
            segment = ONCE;
    CONDES: type    = destructor,
            label   = __DESTRUCTOR_TABLE__,
            count   = __DESTRUCTOR_COUNT__,
            segment = RODATA;
    CONDES: type    = interruptor,
            label   = __INTERRUPTOR_TABLE__,
            count   = __INTERRUPTOR_COUNT__,
            segment = RODATA,
            import  = __CALLIRQ__;
}
```
**c128.cfg - Option `#2`** 
_(1K common RAM)_

```ini
SYMBOLS {
    __LOADADDR__:  type = import;
    __EXEHDR__:    type = import;
    __STACKSIZE__: type = weak, value = $0800; # 2k stack
}
MEMORY {
    ZP:                  file = "", define = yes, start = $0002, size = $001A;
    LOADADDR:   file = %O,               start = $1BFF, size = $0002;
    HEADER:         file = %O,               start = $1C01, size = $000C;
    MAIN:             file = %O, define = yes, start = $1C0D, size = $A3F0 - __STACKSIZE__, bank = 99; #  $A3F3
    COMMON:      file = %O, define = yes, start = $0202, size = $00A2;
    BANK1:           file = "bank1seg.prg", define = yes, start = $1C0D, size = $A3F0, bank = 1;
}
SEGMENTS {
    ZEROPAGE: load = ZP,       type = zp;
    LOADADDR: load = LOADADDR, type = ro;
    EXEHDR:   load = HEADER,   type = ro;
    STARTUP:  load = MAIN,     type = ro;
    LOWCODE:  load = MAIN,     type = ro,  optional = yes;
    ONCE:     load = MAIN,     type = ro,  optional = yes;
    CODE:     load = MAIN,     type = ro;
    RODATA:   load = MAIN,     type = ro;
    DATA:     load = MAIN,     type = rw;
    INIT:     load = MAIN,     type = bss;
    BSS:      load = MAIN,     type = bss, define   = yes;

#The following segment common memory is added for bank switching code
     BANKING:    load = COMMON,   type = ro,  define = yes,  optional = yes;
     BANK1SEG:  load = BANK1,   type = ro,  define = yes,  optional = yes;
}
FEATURES {
    CONDES: type    = constructor,
            label   = __CONSTRUCTOR_TABLE__,
            count   = __CONSTRUCTOR_COUNT__,
            segment = ONCE;
    CONDES: type    = destructor,
            label   = __DESTRUCTOR_TABLE__,
            count   = __DESTRUCTOR_COUNT__,
            segment = RODATA;
    CONDES: type    = interruptor,
            label   = __INTERRUPTOR_TABLE__,
            count   = __INTERRUPTOR_COUNT__,
            segment = RODATA,
            import  = __CALLIRQ__;
}
```
**c128.cfg - Option `#3`**
_(4K common RAM, using POKE 54534,5 in crt0.s, and when starting the actual c128 before loading the prg or bload the bank1seg.)_

```ini
SYMBOLS {
    __LOADADDR__:  type = import;
    __EXEHDR__:    type = import;
    __STACKSIZE__: type = weak, value = $0800; # 2k stack
}
MEMORY {
    ZP:                  file = "", define = yes, start = $0002, size = $001A;
    LOADADDR:   file = %O,               start = $1BFF, size = $0002;
    HEADER:         file = %O,               start = $1C01, size = $000C;
    MAIN:             file = %O, define = yes, start = $1C0D, size = $A3F0 - __STACKSIZE__, bank = 99; #  $A3F3
    COMMON:      file = %O, define = yes, start = $0402, size = $03FD;
    BANK1:           file = "bank1seg.prg", define = yes, start = $1C0D, size = $A3F0, bank = 1;
}
SEGMENTS {
    ZEROPAGE: load = ZP,       type = zp;
    LOADADDR: load = LOADADDR, type = ro;
    EXEHDR:   load = HEADER,   type = ro;
    STARTUP:  load = MAIN,     type = ro;
    LOWCODE:  load = MAIN,     type = ro,  optional = yes;
    ONCE:     load = MAIN,     type = ro,  optional = yes;
    CODE:     load = MAIN,     type = ro;
    RODATA:   load = MAIN,     type = ro;
    DATA:     load = MAIN,     type = rw;
    INIT:     load = MAIN,     type = bss;
    BSS:      load = MAIN,     type = bss, define   = yes;

#The following segment common memory is added for bank switching code
     BANKING:    load = COMMON,   type = ro,  define = yes,  optional = yes;
     BANK1SEG:  load = BANK1,   type = ro,  define = yes,  optional = yes;
}
FEATURES {
    CONDES: type    = constructor,
            label   = __CONSTRUCTOR_TABLE__,
            count   = __CONSTRUCTOR_COUNT__,
            segment = ONCE;
    CONDES: type    = destructor,
            label   = __DESTRUCTOR_TABLE__,
            count   = __DESTRUCTOR_COUNT__,
            segment = RODATA;
    CONDES: type    = interruptor,
            label   = __INTERRUPTOR_TABLE__,
            count   = __INTERRUPTOR_COUNT__,
            segment = RODATA,
            import  = __CALLIRQ__;
}
```

**Reply by RetrofanC** [reply]
(Sorry the # in options didn't like the bold and italic ... I forgot to check the preview this time... )-8)

**Comment by mrdudz** [comment]
use markup (see my edit) :)

can you create a simple asm file too that shows the problem?

**Reply by RetrofanC** [reply]
Hi,
Below a part of the asm program, the first rountin in there.

When I use c128.,cfg - option 1 it is not working nor visible at $03E4 which shows just the same values as it does after reboot.
When I use c128.,cfg - option 2 it is not working nor visible at $0200 which shows lots of unrecognizable code.

Now to be fair, as for both options above I check it currently with the built-in monitor. 
So I can image that switching in the bank 15 with all hoops available, Kernel, Basic, Monitor, etc.
That overwrites the $0200 section before I can see it. But, I wouldn't expect it clears out the #03E4 section as well, but it might.

That is why I switched to a 4k common RAM, and use the 40-column screen memory ($0402), as at least I can see what is going on and if switching in bank config 15 does clear that area of common RAM as well.

The result is that I don't see that happening, but neither is the "BANKING" segment and code loaded when I load and run it. 
(At least after loading, I can't find the code at the $0402 location, and only after loading it separate (changing the config file) as "commonram" with `BLOAD"commonram",b1, p1026`, it is available.

kernalback.s
=========================
```asm
.export _kernalback
.importzp tmp1, tmp2

.segment	"BANKING"
.proc _kernalback
            nop                        ; when using file = "commonram" the first two byte seem to disapear so we need at least two NOP's
            nop                        ; when using file = "commonram" the first two byte seem to disapear so we need at least two NOP's
            nop
            pha
            lda tmp1                 ; this tmp1 hold the original bank configuration value 
            sta $ff00
            pla
            rts
.endproc
```
If you say it should work, what I will do is take my config c128.cfg option 1, and try this ASM. 
Then see if it will put the `@` in the middle of the screen. 
As that should work then, right? 

test_common_area.s
=========================
```asm
.export _test_common_area
.importzp tmp1, tmp2

.segment	"BANKING"
.proc _test_common_area
            nop                        ; when using file = "commonram" the first two byte seem to disapear so we need at least two NOP's
            nop                        ; when using file = "commonram" the first two byte seem to disapear so we need at least two NOP's
            nop
            pha
            lda ##00 
            sta $0550
            pla
            rts
.endproc
```

**Reply by greg-king5** [reply]
> **kernalback.s**
> ```assembly
> .export _kernalback
> .importzp tmp1, tmp2
> 
> .segment	"BANKING"
> .proc _kernalback
>         nop  ; when using file = "commonram", the first two bytes seem to disappear, so we need at least two NOPs
>         nop  ; when using file = "commonram", the first two bytes seem to disappear, so we need at least two NOPs
> ```

That's because every file that's loaded must have a load address.  It doesn't matter what kind of file.  It doesn't matter if the address is used or not used.  Some kind of load address **always** must exist!  If it doesn't exist, then KERNAL will take the first two bytes from your file, and use them as the address.

ld65 doesn't make load addresses.  We must do it manually.  That's what the `LOADADDR` segment does in many configure files.
```assembly
.export _kernalback
.importzp tmp1, tmp2

.segment	"BANKING"
        .addr   * + 2  ; load address
.proc _kernalback
        pha
        lda tmp1       ; this tmp1 hold the original bank configuration value
        sta $ff00
        pla
        rts
.endproc
```

**Reply by RetrofanC** [reply]
Understood, makes sense and thanks for the example, appreciated!

**Comment by mrdudz** [comment]
We really need a complete fully working example to work this out... including the commands you are using to assemble and link (your comments about the first two bytes suggest something is wrong there)

**Reply by RetrofanC** [reply]
**c128.cfg**
```ini
SYMBOLS {
    __LOADADDR__:  type = import;
    __EXEHDR__:    type = import;
    __STACKSIZE__: type = weak, value = $0800; # 2k stack
}
MEMORY {
    ZP:       file = "", define = yes, start = $0002, size = $001A;
    LOADADDR: file = %O,               start = $1BFF, size = $0002;
    HEADER:   file = %O,               start = $1C01, size = $000C;
    MAIN:     file = %O, define = yes, start = $1C0D, size = $83F0 - __STACKSIZE__, bank = 99; 
    BANK1:    file = "bank1seg.prg", define = yes, start = $83F1, size = $1FFF, bank = 1;
    COMMON:   file = %O, start = $03E4, size = $001B,define = yes;
}
SEGMENTS {
    ZEROPAGE: load = ZP,       type = zp;
    LOADADDR: load = LOADADDR, type = ro;
    EXEHDR:   load = HEADER,   type = ro;
    STARTUP:  load = MAIN,     type = ro;
    LOWCODE:  load = MAIN,     type = ro,  optional = yes;
    ONCE:     load = MAIN,     type = ro,  optional = yes;
    CODE:     load = MAIN,     type = ro;
    RODATA:   load = MAIN,     type = ro;
    DATA:     load = MAIN,     type = rw;
    INIT:     load = MAIN,     type = bss;
    BSS:      load = MAIN,     type = bss, define   = yes;
    BANKING:   load = COMMON,   type = ro,  define = yes,  optional = yes;
    BANK1SEG:  load = BANK1,   type = ro,  define = yes,  optional = yes;
}
FEATURES {
    CONDES: type    = constructor,
            label   = __CONSTRUCTOR_TABLE__,
            count   = __CONSTRUCTOR_COUNT__,
            segment = ONCE;
    CONDES: type    = destructor,
            label   = __DESTRUCTOR_TABLE__,
            count   = __DESTRUCTOR_COUNT__,
            segment = RODATA;
    CONDES: type    = interruptor,
            label   = __INTERRUPTOR_TABLE__,
            count   = __INTERRUPTOR_COUNT__,
            segment = RODATA,
            import  = __CALLIRQ__;
}
```

**kernalback.s**
```asm
.export _kernalback

.segment	"BANKING"
.proc _kernalback
            nop
            nop
            nop
            pha
            lda #$00
            sta $0550
            pla
            rts
.endproc
```

**kernal.c**
```c
#include <stdio.h>

extern void kernalback(void);

void main (void)
{
printf("Hello World! \n");
kernalback();
}
```
**Compiler cmd**
```bash
cl65 -O -t c128 kernalback.s kernal.c -o kernal.prg -m kernal_map.txt 
```
**C128 Emulator**
Menu Load File
RUN

**Screen Result**
![image](https://user-images.githubusercontent.com/66948775/139076853-d826545f-85e3-4979-b16a-4e618414022b.png)
Because there is no code at location $03e4 and onwards.

When I change the c128.cfg with; 
```ini
COMMON:   file = "commonram",  start = $03E4, size = $001B,define = yes;
```
```sh
cl65 -O -t c128 kernalback.s kernal.c -o kernal.prg -m kernal_map.txt 
```
Then do a `BLOAD"commonram",B0,P996`.
Menu Load "Kernal.prg".
Check first if all is in place with the MONITOR command.
![image](https://user-images.githubusercontent.com/66948775/139077588-9f23b1d9-78b4-48dc-9ac1-fc4212e8deda.png)
After that, do a RUN command in BASIC.
![image](https://user-images.githubusercontent.com/66948775/139077772-6aa452a5-d822-437f-a210-f2ef3046218f.png)
It works as it should. 

So, why with just %O, doesn't it become part of my kernal.prg? 

Does the above full example with all files and commands help?

**Reply by RetrofanC** [reply]
Just to check it is not BASIC or the MONITOR interfering with the results.
If I show the $03e4 values from within the program with a peek:

**kernal.c**
```c
#include <stdio.h>
#include <peekpoke.h>

extern void kernalback(void);

void main (void)
{
int x;
int addr;
int value;


printf("Hello World! \n\n");
for (x=0; x<10;x++)
{
addr = 996 + x;
value = PEEK(addr);
printf("Value of Location %x = %x \n", addr,value);
}
kernalback();
}
```
I am getting zeroes, as well, until the DMA routine starts at $03f0.
![image](https://user-images.githubusercontent.com/66948775/139083332-9d509671-f9c2-4441-a9cb-a0850e115143.png)

**Comment by groessler** [comment]
Can you privide a zip file with everything included, and batch or shell scripts to build?

**Reply by RetrofanC** [reply]
Sure and thanks all for the ongoing support to get to the bottom of this! 

[kernal.zip](https://github.com/cc65/cc65/files/7427506/kernal.zip)

(I commented the extra loop out which showed the values of $03E4 onwards from within the application)

**Reply by RetrofanC** [reply]
Hi, 

Some additional interesting finds; 

When I place my 

COMMON:   file = %O, define = yes, start = $03E4, size = $001B, fill=0;

in the c128.cfg between;  

ZP: and LOADADDR 
or 
LOADADDR and HEADER 

My test routine;
LDA #$00 
STA $0550 

is actually found at memory location $1C01

When I place my 

COMMON:   file = %O, define = yes, start = $03E4, size = $001B, fill=0;

the c128.cfg after MAIN: 

My routine is actually found at $26AA 
(This looks a very strange location "in the middle of nowhere", which I cannot relate to anything)

Two preliminary conclusions based on this; 

So the location of my COMMON: line within the c128.cfg file impacts which memory address is actually used for my routine.
Regardless of the location of the COMMON: line within the c128.cfg file, it ignores in both cases the given start address of $03e4. 

The wondering grows...

**Comment by groessler** [comment]
In your original kernal.zip the content of COMMON is there, right at the end of the file, as set in the cfg file:

```
$ hexdump -C kernal.prg | tail -3
00000a40  03 00 00 00 00 00 00 00  02 24 00 00 00 00 ea ea  |.........$......|
00000a50  ea 48 a9 00 8d 50 05 68  60                       |.H...P.h`|
00000a59
```

**Reply by RetrofanC** [reply]
Thanks! 
Yes that matches with my latest finding that the code is at a very different location as expected.

**Comment by greg-king5** [comment, marked as answer]
You cannot do what you want to do!  Commodore program files don't work that way.  A file is loaded as a _single_ contiguous blob.  If you want to put data at two or more places, then either separate files must be loaded; or, your program manually must copy special segments from the blob over to where you want them to be.  That's why we have the `run=` attribute.  Your configure file would need to have lines that look like these:
```ini
COMMON:   file = "", start = $03E4, size = $001B, bank = 99;
  ...
BANKING:   load = MAIN, run = COMMON, type = ro, define = yes, optional = yes;
```

**Reply by RetrofanC** [reply]
Understood, that makes sense, and explains it all. 

So everything I define in the MEMORY section behind MAIN: will just be put in the actual a.out file at the end of that file.
(As Groelser shows above as well.)
If I put it in the MEMORY section, somewhere in between line items it will add the code in between there.
The reference 'start' and 'length'  are used to create the labels (use memory locations in the code); and with the `define=yes`, I can find back where the code actually is in the c128 memory.

Now, one last snag, unfortunately.
So, I did what you suggested above, and changed the c128.cfg accordingly, and added the memcpy to the .c file. 

Now, compile is OK.
Loading is OK.

However,
when I check memory locations $0278b, there is the 'COMMON' code in 'MAIN' memory -- all good.
![image](https://user-images.githubusercontent.com/66948775/139224701-ff9952f5-def7-433e-92d7-8c673e733c71.png)

When I check memory locations $03e4, there is just the boot code, mostly $00 brk which is all good.

When I run the program, it runs and comes back to basic nicely.

However, 
it didn't copy the COMMON code from $278b -> $$03e4 !!
it did copy code from $278b -> $03e4, but very different than what I put there!! 
![image](https://user-images.githubusercontent.com/66948775/139224859-0a0549fe-f49b-4371-aa64-eb22813c8c58.png)

When I, after that, compare it with the 'current' contents of the location $0278b, it is the same code over here as well. 
![image](https://user-images.githubusercontent.com/66948775/139224949-d727240d-3ec3-4268-93e3-ca7a194d0ce5.png)

So, the memcpy did well in copying the $278b contents into $03e4, but it's no longer the right code. 

So, when the program is run, and before the memcpy can kick in, the contents of that memory location $278b is changed!!

[Q} Any idea how this can happen and how I can make sure the memcpy is done before the code is being changed? 

For completeness, the simple test.c I used:
```c
#include <stdio.h>
#include <string.h>

extern void print_on_screen(void);

void main(void)
{
char *src;
char *dest;

dest=(char *)0x03e4;
src=(char *)0x278b;
memcpy(dest,src,0x16);

printf("Hello World! \n");
}
```

**Reply by greg-king5** [reply]
> So, when the program is run, and before the memcpy can kick in, the contents of that memory location $278b is changed!!

Quick fix:  Move your `BANKING` line up above your `INIT` line.

The reason:
Segments that are marked `type=bss` start their lives with no data in them (ld65 won't let you put data into them).  They begin as holes in the memory map.  ld65 doesn't write those segments into the output program files (the holes shrink to nothing in the files).  Therefore, any segments that come after those `type=bss` ones are loaded into the wrong places.  When a program starts, data is written to variables in `INIT` and `BSS`.  Your `BANKING` segment got "nuked"!

We must remember two rules:

1. Group all `type=bss` lines together at the end of the list of segments.
2. `BSS` must be the last segment.

**Reply by RetrofanC** [reply]
Hi Greg, 

Brilliant and I am happy bunny as all works as it should now for sections $03e4, $0200 in regular common ram etc.
Even better it increases the insights in how cc65 works and I have to say its a magical tool chain if you know the ins and outs as well.

When I have my full project working I might write-up my finding and we can add it to documentation or a how to. 

One last question on this: 

[Q#1] Is there any way in which I can 'release' the 'stub' code again and free up that memory again? 

As I have double code if I use this mechanism (once in MAIN and once in common ram after copying it there)

Surely, it's no biggy as the code in common ram is rather little and with this all working I open up 4 banks of ram with 64k (or 256k ram)
But if there is an easy trick to free up or re-use memory that would be great. 

Per example in the end game I will be using 'overlays' as well. 
So it would be nice if the memory I initially need for filling the common ram.
Can come available again as free memory (or over-writable) after all is 'initialized'  
As I will also be using 'overlay' code in my program and would be sweet if I can use that space again to put my overlay code.

Which might be a bit tricky as you mentioned BSS code always comes last.
[Q#2] Is that merely for the cc65 tool or is that actually also 'last; in the real memory ?

Thanks!

**Reply by RetrofanC** [reply]
Hi.
As mentioned, it is working better and better on a functional level. 
However, currently I have to manual change the value of the starting address all the time. 
```c
dest=(char *)0x0402;
src=(char *)0x27ff; <-- this I have to change manually every time I change something in the code, as that address moves.
memcpy(dest,src,0x85);
```
I want to get that done automatic by using the Linker Generated 'labels', as described in the cc65 documentation.
However, I can't seem to find those labels, nor get them to work!
(They don't show in the 'MAP' file, as the `"__BSS_START__"` does.
Should they be available in the 'MAP' file?
Or, are they 'hidden' somehow?)

I use the suggested new segment and memory at the right locations in the file, as suggested in the reply above.
Which makes my code work nicely; only, I can't find the suggested three 'labels' indicating the START, LOAD, and RUN locations.
```ini
COMMON:   file = "", start = $03E4, size = $001B, bank = 99;
  ...
BANKING:   load = MAIN, run = COMMON, type = ro, define = yes, optional = yes;
```

The questions; 

[Q1] This is how the labels should be called right?;
```
__BANKING_START__
__BANKING_LOAD__
__BANKING_RUN__
```

[Q2] To use them, I should put this in my banking.c code right?
```asm
.import   __BANKING_LOAD__, __BANKING_RUN__, __BANKING_START__
```

[Q3] Where should I put the ".export "? In which file are they generated by the linker? 
Perhaps in the c128.cfg file?
Or in one of the banking.s files perhaps where I use the .segment "BANKING" ? 

[Q4] How should I use them in my code?
Normally, I have to omit the underscore predecessor " _ " which is used in the .s file when I use them in my .c code.
But, how should that work here, as there are several underscores used?
```c
printf("\nset irq =%04x \n",__BANKING_RUN__); <-- keep all underscores intact.
printf("\nset irq =%04x \n",_BANKING_RUN__);  <-- remove only one underscore from the beginning...
```
Or, a different way all together? 

So, when using it eventually in my code, how should it look, like this? 
```c
dest=(char *)0x0402;
src=(char *)_BANKING_LOAD__; <-- with one less underscore
memcpy(dest,src,0x85);
```
All help is welcome, and to be sure I do spend a lot of time trying to find the answers on the internet and in the documentation.
But also, in this case, I can find lots of related clues; but, I can't find the full story/example which binds all together, so I can make it work.

Thanks in advance!

**Reply by greg-king5** [reply]
[A1] The normal map file shows only labels that were used by the program.  But, you can include the option `-vm` on your command line.  The map file becomes verbose; it shows all labels that were exported by the program and the linker.

[A2] `.import` (and `.export`) are for Assembly files.  C source files use
```c
extern void _BANKING_LOAD__[], _BANKING_RUN__[], _BANKING_SIZE__[];
```

[A3] The `define=yes` ld65 attribute does the exporting.

[A4]
```c
memcpy(_BANKING_RUN__, _BANKING_LOAD__, (size_t)_BANKING_SIZE__);
```

**Reply by RetrofanC** [reply]
Thanks!! 
All clear and works like a charm!! (-8

**Comment by polluks** [comment]
Indeed:
PRG load address
SEQ no load address

**Reply by mrdudz** [reply]
> Indeed: PRG load address SEQ no load address

please lets not spread this misinformation - try SAVE"foo,s,w",8 - SEQ or PRG is totally unrelated here

**Comment by polluks** [comment]
This is cheating :-P

2021-10-27 23:55 GMT+02:00, Bob Andrews ***@***.***>:
